src/middlewared/middlewared/plugins/reporting/netdata_web.py

```diff
@@ -0,0 +1,86 @@
+import os
+import shutil
+import threading
+import time
+
+from middlewared.schema import accepts, returns, Str
+from middlewared.service import cli_private, job, pass_app, periodic, private, CallError, Service
+from middlewared.utils import MIDDLEWARE_RUN_DIR
+from middlewared.utils.crypto import generate_string
+from passlib.apache import HtpasswdFile
+
+
+BASIC_FILE = f'{MIDDLEWARE_RUN_DIR}/netdata-basic'
+HTPASSWD_LOCK = threading.Lock()
+
+
+class ReportingService(Service):
+
+    @private
+    async def netdataweb_basic_file(self):
+        return BASIC_FILE
+
+    @cli_private
+    @accepts(roles=['READONLY_ADMIN'])
+    @returns(Str('password'))
+    @pass_app()
+    def netdataweb_generate_password(self, app):
+        """
+        Generate a password to access netdata web.
+        That password will be stored in htpasswd format for HTTP Basic access.
+        """
+        # Password schema is not used here because for READONLY_ADMIN
+        # will make it return "******" instead, breaking this method for that role.
+        if app and app.authenticated_credentials.is_user_session:
+            authenticated_user = app.authenticated_credentials.user['username']
+        else:
+            raise CallError('This method needs to be called from an authenticated user only.')
+
+        if not os.path.exists(BASIC_FILE):
+            with open(os.open(BASIC_FILE, flags=os.O_CREAT, mode=0o640)):
+                shutil.chown(BASIC_FILE, 'root', 'www-data')
+
+        with HTPASSWD_LOCK:
+            ht = HtpasswdFile(BASIC_FILE, autosave=True, default_scheme='bcrypt')
+            password = generate_string(16, punctuation_chars=True)
+            ht.set_password(authenticated_user, password)
```

Comment:
I think at a minimum we should check whether the user already exists in the htpasswd file, and generate a debug message so that we have some insight into a situation that may break someone's netdata access.

Reply:
Added a warning but can change to a debug?